runtime/vm/CRIUHelpers.cpp

```diff
@@ -389,4 +402,79 @@ runInternalJVMRestoreHooks(J9VMThread *currentThread)
 	return result;
 }
 
+BOOLEAN
+runDelayedLockRelatedOperations(J9VMThread *currentThread)
+{
+	J9JavaVM *vm = currentThread->javaVM;
+	J9InternalVMFunctions* vmFuncs = vm->internalVMFunctions;
+	J9DelayedLockingOpertionsRecord *delayedLockingOperation = (J9DelayedLockingOpertionsRecord*) J9_LINKED_LIST_START_DO(vm->checkpointState.delayedLockingOperationsRoot);
+	J9DelayedLockingOpertionsRecord *lastOperation = NULL;
+	BOOLEAN rc = TRUE;
+
+	Assert_VM_true(vm->checkpointState.checkpointThread == currentThread);
+
+	while (NULL != delayedLockingOperation) {
+		omrthread_monitor_t monitorPtr = NULL;
+		j9object_t instance = J9_JNI_UNWRAP_REFERENCE(delayedLockingOperation->globalObjectRef);
+		if (!VM_ObjectMonitor::inlineFastObjectMonitorEnter(currentThread, instance)) {
+			rc = objectMonitorEnterNonBlocking(currentThread, instance);
+			if (J9_OBJECT_MONITOR_BLOCKING == rc) {
+				/* owned by another thread */
+				Trc_VM_criu_runDelayedLockRelatedOperations_contendedMonitorEnter(currentThread, instance);
+				rc = FALSE;
+				goto done;
+			} else if (J9_OBJECT_MONITOR_ENTER_FAILED(rc)) {
+				/* not possible if the the application was able to call notify earlier */
+				Assert_VM_unreachable();
+			}
+		}
+		if (!VM_ObjectMonitor::getMonitorForNotify(currentThread, instance, &monitorPtr, true)) {
+			if (NULL != monitorPtr) {
+				/* another thread owns the lock, shouldn't be possible */
+				Trc_VM_criu_runDelayedLockRelatedOperations_contendedMonitorEnter2(currentThread, monitorPtr);
+				rc = FALSE;
+				goto done;
+			} else {
+				/* no waiters */
+				goto next;
+			}
+		}
+
+		Trc_VM_criu_runDelayedLockRelatedOperations_runDelayedOperation(currentThread, delayedLockingOperation->operation, instance, monitorPtr);
+
+		switch(delayedLockingOperation->operation) {
+		case J9_SINGLE_THREAD_MODE_OP_NOTIFY:
+			rc = 0 == omrthread_monitor_notify(monitorPtr);
+			break;
+		case J9_SINGLE_THREAD_MODE_OP_NOTIFY_ALL:
+			rc = 0 == omrthread_monitor_notify_all(monitorPtr);
+			break;
+		default:
+			Assert_VM_unreachable();
+			break;
+		}
+
+		if (!rc) {
+			goto done;
+		}
+
+next:
+		if (!VM_ObjectMonitor::inlineFastObjectMonitorExit(currentThread, instance)) {
+			if (0 != objectMonitorExit(currentThread, instance)) {
+				Assert_VM_unreachable();
+			}
+		}
+
+		vmFuncs->j9jni_deleteGlobalRef((JNIEnv*) currentThread, delayedLockingOperation->globalObjectRef, JNI_FALSE);
+		lastOperation = delayedLockingOperation;
```

Comment:
Given this is a cpp file, can we declare `lastOperation` here rather than at the top of the scope?  It makes it easer to reason about the lifetime of the variable
```suggestion
		J9DelayedLockingOpertionsRecord *lastOperation = delayedLockingOperation;
```